### PR TITLE
never_patterns: Parse match arms with no body

### PR DESCRIPTION
**Pull request by Nadrieril**
Never patterns are meant to signal unreachable cases, and thus don't take bodies:
```rust
let ptr: *const Option<!> = ...;
match *ptr {
    None => { foo(); }
    Some(!),
}
```
This PR makes rustc accept the above, and enforces that an arm has a body xor is a never pattern. This affects parsing of match arms even with the feature off, so this is delicate. (Plus this is my first non-trivial change to the parser).

~~The last commit is optional; it introduces a bit of churn to allow the new suggestions to be machine-applicable. There may be a better solution? I'm not sure.~~ EDIT: I removed that commit

r? @compiler-errors

### CODE REVIEW

**Comment by Nadrieril** [comment]
(only the last commit changes to the size of AST and/or HIR nodes)

File: tests/ui/rfcs/rfc-0000-never_patterns/parse.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+fn main() {}
+
+macro_rules! never {
+    () => { ! }
+}
+
+fn parse(x: Void) {
+    match None::<Void> {
+        None => {}
```
Comment (compiler-errors) [inline tests/ui/rfcs/rfc-0000-never_patterns/parse.rs:14]: What happens when we do `! | _`?

e.g.

```
match None::<!> {
  Some(!) | _
```

Comment (compiler-errors) [reply tests/ui/rfcs/rfc-0000-never_patterns/parse.rs:14]: Actually, I misinterpreted what `is_never_pattern` was doing. Okay, that makes sense!

**Comment by calebcartwright** [comment]
@compiler-errors do you think this (or perhaps the associated tracking issue/rfc) would be worth t-style discussion?

**Comment by Mark-Simulacrum** [comment]
> This affects parsing of match arms even with the feature off, so this is delicate.

Is this change FCP'd or nightly-only? It sounds like this is adding new syntax to the language -- is there a lang sign-off on that somewhere?

(I'm surprised this landed without that).

**Comment by Mark-Simulacrum** [comment]
Perf regression looks to be genuine but likely not worth investigating, limited to single primary incr-unchanged scenario.

**Comment by Nadrieril** [comment]
> It sounds like this is adding new syntax to the language -- is there a lang sign-off on that somewhere?

It's not adding syntax to stable rust. What's different is that now the parser accepts match arms with no body and they're rejected later, instead of them being rejected by the parser directly.

**Comment by petrochenkov** [comment]
>What's different is that now the parser accepts match arms with no body and they're rejected later, instead of them being rejected by the parser directly.

The difference is that
```rust
match *ptr {
    None => { foo(); }
    #[cfg(FALSE)]
    Some(!),
}
```
will compile on stable.

Any unstable syntax needs to be feature-gated at parser level using `gated_spans`.

**Comment by Nadrieril** [comment]
Oh shoot yeah I didn't consider that. I knew there would be a subtlety. To be precise,
```rust
match *ptr {
    None => { foo(); }
    #[cfg(FALSE)]
    Some(!),
}
```
does not compile on stable because `!` is correctly gated at parser level: https://github.com/rust-lang/rust/blob/master/tests/ui/feature-gates/feature-gate-never_patterns.rs

However this now compiles on stable:
```rust
match Some(0) {
    None => { foo(); }
    #[cfg(FALSE)]
    Some(_),
}
```
which is much worse.

**Comment by Nadrieril** [comment]
I don't want to gate a token here though, I want to give a user a normal parse error. Is there a way in the parser to know which feature gates are active? This may not even make sense if we aren't done parsing though...

**Comment by Nadrieril** [comment]
Fix is up: https://github.com/rust-lang/rust/pull/118868

**Pull request merged**